src/raft.rs

```diff
@@ -1937,6 +2237,15 @@ impl<T: Storage> Raft<T> {
         self.raft_log.committed = hs.get_commit();
         self.term = hs.get_term();
         self.vote = hs.get_vote();
+        // HACK: Our current Protobuf uses `0` as `None`.
+        // Good thing we know that index 0 of the raft log is always not important in this context.
+        let began_conf_change_at = hs.get_began_conf_change_at();
+        if began_conf_change_at != 0 {
+            let entry = &self.get_store().entries(began_conf_change_at, began_conf_change_at + 1, NO_LIMIT)
```

Comment:
What if the raft log is compacted?
How about remove `began_conf_change_at` from `HardState`, but add `nodes_next_configuration` and `learners_next_configuration` in `ConfState`? So we can construct a ConfChange entry from it.

Reply:
There is an another problem about `began_conf_change_at`: in the current implement snapshots won't touch this field. For snapshot machanism, you can check `Raft::restore_raft` in `raft.rs`.

Reply:
Oh, a good point. I'll investigate this.

Reply:
This should be resolved, a really great catch. Added tests to cover it, too.